Update Repo structure batch 4

chidozieononiwu opened the pull request:
- Updated service location repo root/sdk
- Updated package structure.
- Renamed projects to start with Microsoft.Azure.Batch
- Updated to project properties to match new engineering system.
- Updated TargetFramework to use $(RequiredTargetFramework) i.e. net452, netstandard1.4, net461 and netcoreapp2.1 for test projects.
- Updated CertificateBuilder.cs to run for netcore using BountyCastle instead of Mono.Security.
- Updated CertificateUnitTests.cs
- Disabled CerficaCredentials return on IntegrationTestCommon.cs
- Used Async methods to accomodate .netcore2.1
- Added [LiveTest] Attribute for test requiring environment variable.
- Removed tests in IntegrationJobApplicationPackageReferencesTests.cs due to wrong parameters in BatchServiceClient

chidozieononiwu committed:
Update Repo Structure for Azure Batch DataPlane Library

matthchr commented on sdk/batch/Microsoft.Azure.Batch/tests/IntegrationTestCommon/CertificateBuilder.cs:
```diff
@@ -0,0 +1,100 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
```
Any idea why this file is flagged as a new file and not just a move?

chidozieononiwu replied:
For this class (CertificateBuilder.cs) the certificatebuilder was implemented with [Mono.Security Nuget](https://github.com/Azure/azure-sdk-for-net/blob/2483720ad6ac1208e2044cd46490602353a9001b/src/SDKs/Batch/DataPlane/TestCommon/IntegrationTestCommon/CertificateBuilder.cs#L9) which doesn't work with .NetCore and has been unlisted from Nuget.org. Here I am generating the self signed certificate with BountyCastle libraries. I really need you to take a look at it to be sure it works at it should.

matthchr replied:
Yes, I looked and this looked okay to me. Also, I ran tests which use this (in IntegrationTests) and they passed. So you're good here.

matthchr commented on sdk/batch/Microsoft.Azure.Batch/tests/IntegrationTests/ComputeNodeIntegrationTests.cs:
```diff
@@ -829,7 +839,7 @@ public void ComputeNodeUploadLogs()
                         while (DateTime.UtcNow < timeoutAt)
                         {
-                            blobs = container.ListBlobs();
+                            blobs = await BlobStorageExtensions.ListBlobs(container);
```
Does this exist because storage doesn't have continuation handling in their netcore SDK?

chidozieononiwu replied:
Am not sure what you mean, but I did make use of the continuation token in the implementation of the extension method.[here](https://github.com/Azure/azure-sdk-for-net/pull/6362/files#diff-b2ca8962af117800d7407f8b9ac85c57)

matthchr replied:
I mean that the Azure Storage SDK usually does this for you. The ListBlobs method we were calling previously already did this - I'm wondering why we can't just use an existing ListBlobs from the storage SDK itself rather than write our own extension

chidozieononiwu replied:
The ListBlobs method is not available in .netcore2.1

matthchr commented on sdk/batch/Microsoft.Azure.Batch/tests/IntegrationTests/IntegrationTestUtilities/FileUploadUtilities.cs:
```diff
@@ -252,7 +252,7 @@ private static string ConstructBlobSource(string container, string blob)
         /// <param name="blobUri">blob URI</param>
         /// <returns>the SAS for the container, in full URI format.</returns>
-        private static string CreateContainerWithPolicySASIfNotExist(string account, string key, Uri blobUri, string container, string policy, DateTime start, DateTime end, SharedAccessBlobPermissions permissions)
+        private static async Task<string> CreateContainerWithPolicySASIfNotExist(string account, string key, Uri blobUri, string container, string policy, DateTime start, DateTime end, SharedAccessBlobPermissions permissions)
```
Minor: If this is an async function now, can you add an async suffix here.

chidozieononiwu replied:
Ok, will do.

matthchr commented on sdk/batch/Microsoft.Azure.Batch/tests/IntegrationTests/IntegrationTestUtilities/FileUploadUtilities.cs:
```diff
@@ -308,7 +308,7 @@ private static string CreateContainerWithPolicySASIfNotExist(string account, str
         }
 
-        private static void CreateDefaultBlobContainerAndSASIfNeededReturn(List<IFileStagingProvider> filesToStage, SequentialFileStagingArtifact seqArtifact)
+        private static async Task CreateDefaultBlobContainerAndSASIfNeededReturn(List<IFileStagingProvider> filesToStage, SequentialFileStagingArtifact seqArtifact)
```
Same, async suffix

chidozieononiwu replied:
Ok, will do.

matthchr commented on sdk/batch/Microsoft.Azure.Batch/tests/IntegrationTests/IntegrationTestUtilities/SynchronizationContextHelper.cs:
```diff
@@ -17,7 +16,7 @@ private static void SetSynchronizationContext()
             //Note that XUnit sets the sync context and we want to make sure not to pollute that - this means that
             //we cannot use the Assert.<> functions which take a lamdba and do an await inside XUnit code because it 
             //will cause a deadlock...
-            DispatcherSynchronizationContext synchronizationContext = new DispatcherSynchronizationContext();
```
This was actually intentional -- can you IFDEF this and make it this way in netfx, and take your change in netcore only (for now at least)?

chidozieononiwu replied:
Ok, will do.

matthchr commented on sdk/batch/Microsoft.Azure.Batch/tests/IntegrationTests/Microsoft.Azure.Batch.Integration.Tests.csproj:
```diff
@@ -67,17 +53,14 @@
   <ItemGroup>
     <Reference Include="System.Net" />
     <Reference Include="System.Net.Http" />
-    <Reference Include="System.Windows.Forms" />
```
Keep these in netfx/not in netcore.

matthchr commented on sdk/batch/Microsoft.Azure.Batch/tests/ProtocolTests/IntegrationJobApplicationPackageReferencesTests.cs:
```diff
@@ -0,0 +1,189 @@
+        }
+
+        //[Fact]
+        //[Trait(TestTraits.Duration.TraitName, TestTraits.Duration.Values.MediumDuration)]
```
Was this stuff commented out before you moved it?

matthchr replied:
Looks like it was

chidozieononiwu replied:
No, it wasn't commented out earlier.
I did comment it out. This line
`var client = new BatchServiceClient(new Uri(Url), new BatchSharedKeyCredential(AccountName, AccountKey))` was an error due to wrong arguements.

>Error	CS1503	Argument 1: cannot convert from 'System.Uri' to 'Microsoft.Rest.ServiceClientCredentials'	Microsoft.Azure.Batch.Protocol.Tests (net461), Microsoft.Azure.Batch.Protocol.Tests (netcoreapp2.1)	

> Error	CS1503	Argument 2: cannot convert from 'Microsoft.Azure.Batch.Protocol.BatchSharedKeyCredential' to 'System.Net.Http.DelegatingHandler'	Microsoft.Azure.Batch.Protocol.Tests (net461), Microsoft.Azure.Batch.Protocol.Tests (netcoreapp2.1)

matthchr commented on sdk/batch/Microsoft.Azure.Batch/tests/ProtocolTests/IntegrationJobApplicationPackageReferencesTests.cs:
```diff
@@ -0,0 +1,189 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
```
Honestly if it's easier you can just delete this entire ProtocolTests project. It's all live tests though so if you don't you should flag it as such.

If you don't delete it I will shortly.

chidozieononiwu replied:
Ok

matthchr reviewed (changes requested):
Took a look, it looks mostly ok, had a few comments.

chidozieononiwu commented:
@Azure/azure-sdk-eng

chidozieononiwu commented on eng/.docsettings.yml:
```diff
@@ -61,6 +61,8 @@ known_content_issues:
   - ['sdk/storage/Azure.Storage.Common/swagger/Generator/readme.md','azure-sdk-tools/issues/42']
   - ['sdk/cognitiveservices/Language.TextAnalytics/src/Readme.md','#5499']
   - ['sdk/cognitiveservices/Personalizer/src/Readme.md','#5499']
+  - ['sdk/batch/Microsoft.Azure.Batch/README.md','#5499']
```
There is a Doc Warden script that verifies ReadMes' and enforces a certain set of minimum requirements. This line is added to skip that check while referencing the issue that describes it. So basically the ReadMe will have to be updated before this line can be removed else the build will break in CI. In this case the error would be 

> 1 readme has at least one missing required section.
    \sdk\batch\Microsoft.Azure.Batch\README.md is missing headers with patterns:
    * Azure .+ client library for .NET
    * ^Getting started$
    * ^Key concepts$
    * ^Examples$
    * ^Troubleshooting$
    * ^Next steps$
    * ^Contributing$
    For a rundown on what you need to do to resolve this breaking issue ASAP, check out 
    aka.ms/azure-sdk-analyze-failed

chidozieononiwu commented on sdk/batch/Microsoft.Azure.Batch/src/Microsoft.Azure.Batch.csproj:
```diff
@@ -0,0 +1,56 @@
+﻿<Project Sdk="Microsoft.NET.Sdk">
+  <PropertyGroup>
+    <Description>This client library provides access to the Microsoft Azure Batch service.</Description>
+    <VersionPrefix>10.0.0</VersionPrefix>
```
I did get the latest changes. I think I used the Assembly Version from [AssemblyAttributes.cs](https://github.com/Azure/azure-sdk-for-net/blob/2483720ad6ac1208e2044cd46490602353a9001b/src/SDKs/Batch/DataPlane/Azure.Batch/AssemblyAttributes.cs#L12) Should'nt they be the same?

matthchr replied:
They probably _should_ be the same. If they weren't though use the later one.

chidozieononiwu commented on sdk/batch/Microsoft.Azure.Batch/src/Microsoft.Azure.Batch.csproj:
```diff
@@ -0,0 +1,56 @@
+  </ItemGroup>
+
+  <PropertyGroup>
+    <AzureApiTag>Batch_2019-04-01;</AzureApiTag>
```
That's from [AzSdk.RP.props](https://github.com/Azure/azure-sdk-for-net/blob/2483720ad6ac1208e2044cd46490602353a9001b/src/SDKs/Batch/DataPlane/AzSdk.RP.props#L4) I just moved it into the batch file.

matthchr replied:
Why is the data plane SDK inheriting anything from AzSdk.RP.props? It's not an RP?

This value is wrong so I don't think we should have it here - can we just remove it?

chidozieononiwu commented on sdk/batch/Microsoft.Azure.Batch/tests/IntegrationTestCommon/IntegrationTestCommon.cs:
```diff
@@ -100,8 +99,11 @@ private static CertificateCredentials GetBatchTestTenantCloudCredentials()
             }
-
+#if FullNetFx
             return new CertificateCredentials(certificate2);
```
CertificateCredentials does not work with .NetCore2.1 can you suggest how to get around this?

matthchr replied:
Do you know what the alternative is for CertificateCredentials in netcore? CertificateCredentials AFAIK comes from the Azure SDK so presumably there's some replacement for it?

chidozieononiwu replied:
I haven't found an alternative foe that.

weshaggard commented on sdk/batch/Microsoft.Azure.Batch/Tools/CodeGenerationLibrary/CodeGenerationLibrary.csproj:
```diff
@@ -1,18 +1,17 @@
     <!--After fixing all warnings, enable this option-->
     <TreatWarningsAsErrors>false</TreatWarningsAsErrors>
-    <AssemblyName>CodeGenerationLibrary</AssemblyName>
+    <AssemblyTitle>CodeGenerationLibrary</AssemblyTitle>
```
Why this change? I wouldn't expect you need AssemblyName or AssemblyTitle.

chidozieononiwu replied:
Why would it not be needed? Is it because it is not a shipped package?

weshaggard replied:
It isn't needed for the same reason AssemblyName isn't needed because it is redundant and defaults from the AssemblyName which defaults from the Project Name, https://github.com/dotnet/sdk/blob/2eb6c546931b5bcb92cd3128b93932a980553ea1/src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.DefaultAssemblyInfo.targets#L23.

chidozieononiwu replied:
Thanks

weshaggard commented on sdk/batch/Microsoft.Azure.Batch/Tools/CodeGenerationLibrary/CodeGenerationLibrary.csproj:
```diff
@@ -1,18 +1,17 @@
-    <PackageId>CodeGenerationLibrary</PackageId>
     <Description>Azure Batch object model code generation library</Description>
     <!--After fixing all warnings, enable this option-->
     <TreatWarningsAsErrors>false</TreatWarningsAsErrors>
```
Looks like this is set to false here and then to true a few lines down.

weshaggard commented on sdk/batch/Microsoft.Azure.Batch/Tools/ConfigureAwaitAnalyzer/ConfigureAwaitAnalyzer.csproj:
```diff
@@ -1,14 +1,15 @@
+    <AssemblyTitle>ConfigureAwaitAnalyzer</AssemblyTitle>
     <SignAssembly>true</SignAssembly>
     <DelaySign>true</DelaySign>
+    <AssemblyOriginatorKeyFile>$(RepoEngPath)\MSSharedLibKey.snk</AssemblyOriginatorKeyFile>
```
This shouldn't be needed.

weshaggard commented on sdk/batch/Microsoft.Azure.Batch/Tools/ConfigureAwaitAnalyzer/ConfigureAwaitAnalyzer.csproj:
```diff
@@ -1,14 +1,15 @@
     <DelaySign>true</DelaySign>
+    <AssemblyOriginatorKeyFile>$(RepoEngPath)\MSSharedLibKey.snk</AssemblyOriginatorKeyFile>
     <RestorePackagesPath>$(LibraryNugetPackageFolder)</RestorePackagesPath>
+    <GenerateAssemblyInformationalVersionAttribute>false</GenerateAssemblyInformationalVersionAttribute>
```
Why are you disabling the assembly information attribute?

weshaggard commented on sdk/batch/Microsoft.Azure.Batch/Tools/ConfigureAwaitAnalyzerTests/ConfigureAwaitAnalyzer.Tests.csproj:
```diff
@@ -10,21 +8,20 @@
     <SignAssembly>true</SignAssembly>
     <DelaySign>true</DelaySign>
-    <AssemblyOriginatorKeyFile>$(LibraryToolsFolder)\MSSharedLibKey.snk</AssemblyOriginatorKeyFile>
+    <AssemblyOriginatorKeyFile>$(RepoEngPath)\MSSharedLibKey.snk</AssemblyOriginatorKeyFile>
```
Shouldn't be necessary.

weshaggard commented on sdk/batch/Microsoft.Azure.Batch/src/AssemblyAttributes.cs:
```diff
@@ -0,0 +1,14 @@
+using System.Runtime.InteropServices;
+
+
+#if CODESIGN || TESTSIGN
```
We always sign now so this shouldn't be needed.

chidozieononiwu commented on sdk/batch/Microsoft.Azure.Batch/Tools/CodeGenerationLibrary/CodeGenerationLibrary.csproj:
```diff
@@ -1,18 +1,17 @@
     <TreatWarningsAsErrors>false</TreatWarningsAsErrors>
-    <AssemblyName>CodeGenerationLibrary</AssemblyName>
+    <AssemblyTitle>CodeGenerationLibrary</AssemblyTitle>
     <RestorePackagesPath>$(LibraryNugetPackageFolder)</RestorePackagesPath>
```
@weshaggard Does this apply in for dataplane?

weshaggard replied:
I don't believe RestorePackagesPath applies.

weshaggard replied:
That said however I'm not entirely sure what it is being used for, it might be getting used by something other then nuget.

weshaggard commented:
/azp run net - client

chidozieononiwu committed:
Added ci.yml for batch service

chidozieononiwu commented:
/azp run net - client

chidozieononiwu commented:
/azp run net - batch - ci

chidozieononiwu commented:
/azp run net - batch - ci

chidozieononiwu commented:
/azp run net - batch - ci

chidozieononiwu commented:
/azp run net - batch - ci

chidozieononiwu commented:
/azp run net - batch - ci

chidozieononiwu commented:
/azp run net - batch - ci

chidozieononiwu commented:
/azp run net - batch - ci

chidozieononiwu commented:
/azp run net - batch - ci

chidozieononiwu commented:
/azp run net - batch - ci

chidozieononiwu committed:
Added preperocessor directive to run test only on windows

chidozieononiwu committed:
Fixed Unit test error

chidozieononiwu commented:
/azp run net - batch - ci

weshaggard commented:
Looks like there are some warnings about test results because some of the test projects don't actually contain tests:

https://dev.azure.com/azure-sdk/public/_build/results?buildId=38551&view=logs - PublishTestResults step

Windows:
```
No Result Found to Publish 'D:\a\1\s\sdk\batch\Microsoft.Azure.Batch\tests\IntegrationTestCommon\TestResults\VssAdministrator_fv-az639_2019-05-25_01_57_42_499.trx'.
No Result Found to Publish 'D:\a\1\s\sdk\batch\Microsoft.Azure.Batch\tests\IntegrationTestCommon\TestResults\VssAdministrator_fv-az639_2019-05-25_01_57_57_715.trx'.
No Result Found to Publish 'D:\a\1\s\sdk\batch\Microsoft.Azure.Batch\tests\ProtocolTests\TestResults\VssAdministrator_fv-az639_2019-05-25_01_58_10_982.trx'.
No Result Found to Publish 'D:\a\1\s\sdk\batch\Microsoft.Azure.Batch\tests\ProtocolTests\TestResults\VssAdministrator_fv-az639_2019-05-25_01_58_14_742.trx'.
No Result Found to Publish 'D:\a\1\s\sdk\batch\Microsoft.Azure.Batch\tests\TestCommon\TestResults\VssAdministrator_fv-az639_2019-05-25_01_58_10_550.trx'.
No Result Found to Publish 'D:\a\1\s\sdk\batch\Microsoft.Azure.Batch\tests\TestCommon\TestResults\VssAdministrator_fv-az639_2019-05-25_01_58_12_335.trx'.
```

The linux and Mac give this warning instead:

```
##[warning]Invalid results file. Make sure the result format of the file '/home/vsts/work/1/s/sdk/batch/Microsoft.Azure.Batch/tests/IntegrationTestCommon/TestResults/_fv-az644_2019-05-25_01_56_30_382.trx' matches 'VSTest' test results format.
##[warning]Invalid results file. Make sure the result format of the file '/home/vsts/work/1/s/sdk/batch/Microsoft.Azure.Batch/tests/ProtocolTests/TestResults/_fv-az644_2019-05-25_01_56_41_929.trx' matches 'VSTest' test results format.
##[warning]Invalid results file. Make sure the result format of the file '/home/vsts/work/1/s/sdk/batch/Microsoft.Azure.Batch/tests/TestCommon/TestResults/_fv-az644_2019-05-25_01_56_33_026.trx' matches 'VSTest' test results format.
```

However I think they stem from the same issue which is there are not test results for those projects.  We should either rename the projects or set IsTestProject=false (similar to     <!-- Disable running of tests for test helper libraries -->
    <IsTestProject Condition="'$(IsTestSupportProject)' == 'true'">false</IsTestProject>) in the project to get them not to run tests and report no results.

chidozieononiwu committed:
Removed Protocol.Test Project, Renamed Common Test Projects

chidozieononiwu commented:
/azp run net - batch - ci

chidozieononiwu commented:
> Looks like there are some warnings about test results because some of the test projects don't actually contain tests:
> 
> https://dev.azure.com/azure-sdk/public/_build/results?buildId=38551&view=logs - PublishTestResults step
> 
> Windows:
> 
> ```
> No Result Found to Publish 'D:\a\1\s\sdk\batch\Microsoft.Azure.Batch\tests\IntegrationTestCommon\TestResults\VssAdministrator_fv-az639_2019-05-25_01_57_42_499.trx'.
> No Result Found to Publish 'D:\a\1\s\sdk\batch\Microsoft.Azure.Batch\tests\IntegrationTestCommon\TestResults\VssAdministrator_fv-az639_2019-05-25_01_57_57_715.trx'.
> No Result Found to Publish 'D:\a\1\s\sdk\batch\Microsoft.Azure.Batch\tests\ProtocolTests\TestResults\VssAdministrator_fv-az639_2019-05-25_01_58_10_982.trx'.
> No Result Found to Publish 'D:\a\1\s\sdk\batch\Microsoft.Azure.Batch\tests\ProtocolTests\TestResults\VssAdministrator_fv-az639_2019-05-25_01_58_14_742.trx'.
> No Result Found to Publish 'D:\a\1\s\sdk\batch\Microsoft.Azure.Batch\tests\TestCommon\TestResults\VssAdministrator_fv-az639_2019-05-25_01_58_10_550.trx'.
> No Result Found to Publish 'D:\a\1\s\sdk\batch\Microsoft.Azure.Batch\tests\TestCommon\TestResults\VssAdministrator_fv-az639_2019-05-25_01_58_12_335.trx'.
> ```
> 
> The linux and Mac give this warning instead:
> 
> ```
> ##[warning]Invalid results file. Make sure the result format of the file '/home/vsts/work/1/s/sdk/batch/Microsoft.Azure.Batch/tests/IntegrationTestCommon/TestResults/_fv-az644_2019-05-25_01_56_30_382.trx' matches 'VSTest' test results format.
> ##[warning]Invalid results file. Make sure the result format of the file '/home/vsts/work/1/s/sdk/batch/Microsoft.Azure.Batch/tests/ProtocolTests/TestResults/_fv-az644_2019-05-25_01_56_41_929.trx' matches 'VSTest' test results format.
> ##[warning]Invalid results file. Make sure the result format of the file '/home/vsts/work/1/s/sdk/batch/Microsoft.Azure.Batch/tests/TestCommon/TestResults/_fv-az644_2019-05-25_01_56_33_026.trx' matches 'VSTest' test results format.
> ```
> 
> However I think they stem from the same issue which is there are not test results for those projects. We should either rename the projects or set IsTestProject=false (similar to https://github.com/Azure/azure-sdk-for-net/blob/master/eng/Directory.Build.Data.targets#L31) in the project to get them not to run tests and report no results.

I have fixed the be renaming the common tests projects. I also deleted the protocol test as @matthchr requested. I guess this is ready to merge now.

matthchr reviewed (approved):
Looks good now
